src/validate.rs

```diff
@@ -1,36 +1,110 @@
-use cketh_common::eth_rpc::ValidationError;
-use ic_cdk::api::management_canister::http_request::HttpHeader;
+use crate::{
+    constants::{CONTENT_TYPE_HEADER, SERVICE_HOSTS_BLOCKLIST, VALID_API_KEY_CHARS},
+    types::ConstHeader,
+    util::hostname_from_url,
+};
 
-use crate::constants::{CONTENT_TYPE_HEADER, SERVICE_HOSTS_BLOCKLIST};
-
-pub fn validate_hostname(hostname: &str) -> Result<(), ValidationError> {
+pub fn validate_hostname(hostname: &str) -> Result<(), &'static str> {
     if SERVICE_HOSTS_BLOCKLIST.contains(&hostname) {
-        Err(ValidationError::HostNotAllowed(hostname.to_string()))
+        Err("Hostname not allowed")
     } else {
         Ok(())
     }
 }
 
-pub fn validate_credential_path(credential_path: &str) -> Result<(), ValidationError> {
-    if !(credential_path.is_empty()
-        || credential_path.starts_with('/')
-        || credential_path.starts_with('?'))
+pub fn validate_url_pattern(url_pattern: &str) -> Result<(), &'static str> {
+    validate_hostname(&hostname_from_url(url_pattern).ok_or("Invalid hostname in URL")?)
+}
+
+pub fn validate_header_patterns(header_patterns: &[ConstHeader]) -> Result<(), &'static str> {
+    if header_patterns
+        .iter()
+        .any(|ConstHeader { name, .. }| *name == CONTENT_TYPE_HEADER)
```

Comment:
note that HTTP headers keys are case-insensitive, see https://datatracker.ietf.org/doc/html/rfc7230#section-3.2, meaning that you would need first to normalise `name` and `CONTENT_TYPE_HEADER` for this check to be effective (e.g., everything lower-case). 

Since all headers name are hard-coded, do we even need this check? I have  the impression it would be a better fit to have such a check as a unit test, to ensure we didn't do something stupid. WDYT?

Reply:
We can probably remove this validation logic. If someone is passing a `Content-Type` header for a custom provider, there is most likely a good reason they're doing that. 

Replaced the header validation with an equivalent unit test for now.

Reply:
> If someone is passing a Content-Type header for a custom provider

ha true, I forgot about `RpcServices::Custom`. 

> If someone is passing a Content-Type header for a custom provider, there is most likely a good reason they're doing that.

Yes, it's a usual source of bugs if some intermediate process is overwriting provided headers. I would add the content-type header only if it was not present in the request.

Reply:
Updated the HTTP request logic so that it's possible to pass a custom content-type header.